fix(40320): Better errors when using properties/methods from newer versions of ECMAScript

Vboivin opened the pull request:
Fixes #40320 and #40319

# New Error Messages
## Error message when properties/methods are missing due to lib configuration
- Added a new error message when properties/methods from newer versions cannot be found because of the current lib configuration
- New message template: 
```js
"Property '{0}' does not exist on type '{1}'. Do you need to change your target library? Try changing the `lib` compiler option to '{2}' or later."
```

## Error message when the name cannot be found due to lib configuration
- Added new identifiers to the already existing function getCannotFindNameDiagnosticForName for es2017, es2018 and es2020:
- - SharedArrayBuffer, Atomics, AsyncIterable, AsyncIterableIterator, AsyncGenerator, AsyncGeneratorFunction, BigInt, BigInt64Array and BigUint64Array
- Modified the message template to be able to pass the suggested lib configuration
- New message template: 
```js
"Cannot find name '{0}'. Do you need to change your target library? Try changing the `lib` compiler option to '{1}' or later."
```
## Preview
![new_error_messages](https://user-images.githubusercontent.com/28627982/93686988-14d1a000-fa88-11ea-959f-c32d04770456.png)

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -2297,6 +2301,10 @@
         "category": "Error",
         "code": 2578
     },
+    "Cannot find name '{0}'. Do you need to change your target library? Try changing the `lib` compiler option to es2017 or later.": {
+        "category": "Error",
+        "code": 2579
+    },
```
Can you remove the new messages and switch the old one to accept a 2nd template slot?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25844,7 +25895,22 @@ namespace ts {
+                return getSpellingSuggestion(name, symbols, getCandidateName);
+            }
+
+            function isSpellingSuggestionEdgeCase(possibleEdgeCase: string) {
+                /**
+                 * Without this check, the user could be returned an unhelpful error message.
+                 * One case would be with getOwnPropertyDescriptor and getOwnPropertyDescriptors.
+                 * In this case, the user would be suggested to target getOwnPropertyDescriptor.
+                 * If he wanted to use the es2017 feature, he should be suggested to change his target lib to es2017.
+                 *
+                 */
+                const SPELLING_EDGE_CASES = ["getOwnPropertyDescriptors"];
+                return SPELLING_EDGE_CASES.includes(possibleEdgeCase);
+            }
+
```
1. You don't need a whole function for this
2. I think you can move this check into `getSuggestedSymbolForNonexistentProperty` - not sure what @sandersn thinks

Vboivin replied:
1. Fair enough
2. I have tried implementing it in my own function, but my diagnostic would get replaced by the one generated by getSpellingSuggestionForName

Vboivin replied:
I got rid of this check entirely in my upcoming code. I feel like this is out of scope for this PR and this should be worked as a separate bug fix for `getSpellingSuggestionForName`

> Can you remove the new messages and switch the old one to accept a 2nd template slot?

I found many more edge cases when I removed the new messages and switched the old one to accept a 2nd template slot for the lib version. 

For example, all the Math functions included in es2015 such as sinh, cosh, or all other functions one letter away from an existing function would not return the proper error message if they were not all included inside of `isSpellingSuggestionEdgeCase`. 

Instead of returning `... Try changing the lib compiler to  '{1}' or later`, it would return `sinh does not exist, did you mean sin?` which might be confusing for someone who was trying to use sinh but is not using the correct configuration.

Of course, once #40320 is fixed, I am willing to work on a fix for this problem.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25756,6 +25776,37 @@ namespace ts {
+            }
+
+            function generalizeTypeString(typeString: string){
+                if (typeString.includes("[]")) {
```
You really want to see if you have a single TypeReference and get the text of the name of its declaration.

Vboivin replied:
This makes a lot more sense, thanks

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25756,6 +25776,37 @@ namespace ts {
+                if (typeString.includes("[]")) {
+                    return "Array";
+                }
+                else if (typeString.includes("\"\"")) {
```
You want to get the *apparent type* and see if you have type with the name `String`, don't look for quote marks

DanielRosenwasser replied:
I should clarify, you can even check if it's equal to the `globalStringType`

Vboivin replied:
Your first comment was the key! Working on it right now and it is passing the tests without my convoluted implementation with "includes"

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25756,6 +25776,37 @@ namespace ts {
+                else if (typeString.includes("\"\"")) {
+                    return "String";
+                }
+                else if (typeString.includes("Promise")) {
+                    return "Promise";
+                }
```
Same technique with `Array`

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25740,7 +25752,15 @@ namespace ts {
                     else {
-                        errorInfo = chainDiagnosticMessages(elaborateNeverIntersection(errorInfo, containingType), Diagnostics.Property_0_does_not_exist_on_type_1, declarationNameToString(propNode), typeToString(containingType));
+                        const property = declarationNameToString(propNode);
+                        const container = typeToString(containingType);
```
As you can see below, this forces you to reparse. You already have the semantic structure, reuse that.

Vboivin commented:
I think the PR is ready, do you see any major flaws in my solution @DanielRosenwasser

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25740,7 +25758,15 @@ namespace ts {
+                        const container = typeToString(containingType);
+                        const lib = getSuggestedLibForNonexistentProperty(missingProperty, containingType);
+                        if (lib) {
+                            errorInfo = chainDiagnosticMessages(elaborateNeverIntersection(errorInfo, containingType), Diagnostics.Property_0_does_not_exist_on_type_1_Do_you_need_to_change_your_target_library_Try_changing_the_lib_compiler_option_to_2_or_later, missingProperty, container, lib);
```
`elaborateNeverIntersection` should never be combined with this message, it would be extremely confusing if the two interacted.

DanielRosenwasser commented on src/compiler/utilities.ts:
```diff
@@ -562,6 +562,76 @@ namespace ts {
+                Symbol: ["description"]
+            },
+            es2020: {
+                BigInt: [""],
+                BigInt64Array: [""],
+                BigUint64Array: [""],
```
Arrays of empty strings will probably give you some weird false-positives for a test like

```ts
// @target: es5

interface BigInt {}
declare var x: BigInt;
let y = x[""];
```

I would suggest using `emptyArray` instead.

```suggestion
                BigInt: emptyArray,
                BigInt64Array: emptyArray,
                BigUint64Array: emptyArray,
```

DanielRosenwasser reviewed (changes requested):
I think this is *very* close to being ready! Let's keep iterating here.

Vboivin commented:
> I think this is _very_ close to being ready! Let's keep iterating here.

I added your suggested changes, this might be the one, or it might need one more iteration.

Vboivin commented:
Ready for review @DanielRosenwasser

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25868,6 +25894,39 @@ namespace ts {
+                for (const libTarget of libTargets) {
+                    const featuresOfLib = allFeatures[libTarget];
+                    const featuresOfContainingType = featuresOfLib[symbolName(container)];
+                    if (featuresOfContainingType) {
+                        for (const feature of featuresOfContainingType) {
+                            if (feature === missingProperty) {
+                                return libTarget;
+                            }
+                        }
+                    }
```
```suggestion
                        if (featuresOfContainingType !== undefined && contains(featuresOfContainingType, missingProperty)) {
                            return libTarget;
                        }
```

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25868,6 +25894,39 @@ namespace ts {
+
+        function getSuggestedLibForNonExistentProperty(missingProperty: string, containingType: Type) {
+            const container = getApparentType(containingType).symbol;
+            if (container) {
```
Turn this into an early bail-out and return `undefined` so that you can avoid nesting.

DanielRosenwasser commented:
Looks like you'll need to update baselines.

Vboivin commented:
> Looks like you'll need to update baselines.

@DanielRosenwasser Thanks for all the extensive reviews! 

I added the early bail-out and updated the baselines to reflect the changes to the main upstream branch. 

Is there some documentation for all the utility functions such as `contains`? It would be helpful for novice contributors like myself. If there isn't, I will take a deeper look at the utility files.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25877,6 +25903,33 @@ namespace ts {
+
+        function getSuggestedLibForNonExistentProperty(missingProperty: string, containingType: Type) {
+            const container = getApparentType(containingType).symbol;
+            if (container) {
```
```suggestion
            if (!container) {
                return undefined;
            }
```

DanielRosenwasser reviewed (approved):
@sandersn @rbuckton can I get a second reviewer on this?

RyanCavanaugh commented:
These tests 😍
